agent/grpc-external/services/peerstream/subscription_blocking.go

```diff
@@ -93,30 +93,27 @@ func (m *subscriptionManager) syncViaBlockingQuery(
 
 	store := m.getStore()
 
-	for {
+	for ctx.Err() == nil {
 		ws := memdb.NewWatchSet()
 		ws.Add(store.AbandonCh())
 		ws.Add(ctx.Done())
 
-		if result, err := queryFn(ctx, store, ws); err != nil && ctx.Err() == nil {
+		if result, err := queryFn(ctx, store, ws); err != nil {
+			// Return immediately if the context was cancelled.
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				return
+			}
 			logger.Error("failed to sync from query", "error", err)
-
+			waiter.Wait(ctx)
 		} else {
 			select {
 			case <-ctx.Done():
 				return
 			case updateCh <- cache.UpdateEvent{CorrelationID: correlationID, Result: result}:
+				waiter.Reset()
 			}
-
 			// Block for any changes to the state store.
 			ws.WatchCtx(ctx)
 		}
-
-		err := waiter.Wait(ctx)
-		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
-			return
-		} else if err != nil {
-			logger.Error("failed to wait before re-trying sync", "error", err)
```

Comment:
Are we missing the case to emit this logger error?

Reply:
The `waiter.Wait` command can only error whenever the context is cancelled, according to the documentation on the method, so the old `else-if err != nil` should be encountered ever. The new implementation will simply stop and break the loop quietly, just like the old one.

Reply:
Good catch